rust/connlib/tunnel/src/lib.rs

```diff
@@ -68,6 +68,17 @@ pub struct Tunnel<CB: Callbacks, TRoleState, TRole, TId> {
     read_buf: [u8; MAX_UDP_SIZE],
 }
 
+pub struct GatewayTunnel<CB: Callbacks> {
+    callbacks: CallbackErrorFacade<CB>,
+
+    state: GatewayState,
+
+    io: Io,
```

Comment:
I am really liking this:

- `GatewayState is SANS-IO, no side-effects.
- `Io` does all the side-effects

Together they form the tunnel :blush: